repository/src/main/java/org/alfresco/repo/event2/EventGenerator.java

```diff
@@ -368,13 +379,36 @@ private boolean isFilteredChildAssociation(QName childAssocType, String user)
 
     private EventInfo getEventInfo(String user)
     {
-        return new EventInfo().setTimestamp(ZonedDateTime.now())
+        return new EventInfo().setTimestamp(getCurrentTransactionTimestamp())
                               .setId(UUID.randomUUID().toString())
                               .setTxnId(AlfrescoTransactionSupport.getTransactionId())
                               .setPrincipal(user)
                               .setSource(URI.create("/" + descriptorService.getCurrentRepositoryDescriptor().getId()));
     }
 
+    private ZonedDateTime getCurrentTransactionTimestamp()
+    {
+        Long currentTransactionId = nodeDAO.getCurrentTransactionId(false);
+        ZonedDateTime timestamp;
+        if(currentTransactionId != null)
+        {
+            Transaction transaction = nodeDAO.getTxnById(currentTransactionId);
+            Instant commitTimeMs = Instant.ofEpochMilli(transaction.getCommitTimeMs());
+            timestamp = ZonedDateTime.ofInstant(commitTimeMs, ZoneOffset.UTC);
+        }
+        else 
+        {
+            // Sometimes the transaction entry is not created because no nodes were added to the transaction, but we are producing events anywhere.
```

Comment:
If the transaction is not active and nothing was committed to the DB, then there should be no events produced.
I think you can check the transaction state by AlfrescoTransactionSupport.isActualTransactionActive().

Reply:
Totally agree, I already did some experiments about it. Should I do it in this PR or in a separate PR/Issue?

Reply:
Anyway, isActualTransactionActive doesn't work in this case because the transaction it is active, but it is never used so AlfrescoTransactionSupport.getResource(KEY_TRANSACTION) is null.

Reply:
Is it correct that after the changes in EventTransactionListener in this class this code will never be executed? If so, can this if-else block be removed?

Reply:
removed